crates/miden-tx/src/executor/notes_checker.rs

```diff
@@ -0,0 +1,104 @@
+use alloc::vec::Vec;
+
+use miden_lib::{
+    account::interface::NoteAccountCompatibility, note::well_known_note::WellKnownNote,
+};
+use miden_objects::{
+    account::AccountId,
+    block::BlockNumber,
+    note::NoteId,
+    transaction::{InputNote, TransactionArgs},
+};
+use winter_maybe_async::{maybe_async, maybe_await};
+
+use super::{NoteAccountExecution, TransactionExecutor, TransactionExecutorError};
+
+/// This struct performs input notes check against provided target account.
+///
+/// The check is performed using the [NotesChecker::check_notes_consumability] procedure.
+/// Essentially runs the transaction to make sure that provided input notes could be consumed by the
+/// account.
+pub struct NotesChecker {
+    account_id: AccountId,
+    notes: Vec<InputNote>,
+}
```

Comment:
I would probably rename this into something like `NoteExecutionChecker` or `NoteConsumptionChecker`.

Also, should we flip this around so that `NoteConsumptionChecker` looks something like:
```rust
pub struct NoteConsumptionChecker<'a>(&'a TransactionExecutor);

pub fn check_notes_consumability(
    &self,
    account_id: AccountId,
    block_ref: BlockNumber,
    notes: Vec<Note>,
    tx_args: TransactionArgs,
) -> Result<NoteAccountExecution, TransactionExecutorError> {
    ...
}
```